### DISCUSSION TITLE
TIAN BMS

### DISCUSSION THREAD

**Discussion by ghost** [Louisvdw/dbus-serialbattery#132] [General]
Hi Louis, it appears as if the revov.py is not included in the “latest” version of the package and thus not installed as part of the “wget https://raw.githubusercontent.com/Louisvdw/dbus-serialbattery/master/etc/dbus-serialbattery/installrelease.sh”. Do I have to create a revov.py file in the /data/etc/dbus-serialbattery/ folder, or what else must I do to get the system to work correctly with the Revov batteries and the TIAN BMS?
![Putty Image](https://user-images.githubusercontent.com/24652007/174016514-f369275e-ae09-4242-8971-ddcd167260a7.jpg)

**Comment by Louisvdw** [comment]
Hi @bertiebez 
I also picked that up. It is already fixed and is part of the new build v0.12
Just run installrelease.sh again to install it

**Reply by ghost** [reply]
Thanks Louis, I have done that, and I now notice that both the revov.py and revov.pyc appears in the /opt/victronenergy/dbus-serialbattery folder, but I still don’t see the SerialBattery in the Device list on the Cerbo GX. I am using the Victron RS485 to USB cable and it appears as if it reads some data (I can see the LED flashing) – also see the attached image.
![Current Data](https://user-images.githubusercontent.com/24652007/174034898-34b6ebf6-3744-4937-852f-5dd099765352.jpg)

**Reply by Louisvdw** [reply]
can you run on your GX using SSH the command
`sh /data/etc/dbus-serialbattery/reinstalllocal.sh`
See if the no service template error goes away after that.

**Reply by ghost** [reply]
I rand the command in your response and got the same “errors” before and after. Here is a copy of the data from the log file /data/log/dbus-serialbattery.ttyUSB0/current and an image of when I ran the “sh /data/etc/dbus-serialbattery/reinstalllocal.sh”.
![2022-06-16_12-41](https://user-images.githubusercontent.com/24652007/174053807-45cf5f69-5cc0-4c72-9020-449d010bc9cb.jpg)
[Current Data Output.txt](https://github.com/Louisvdw/dbus-serialbattery/files/8918130/Current.Data.Output.txt)

**Reply by Louisvdw** [reply]
For some reason it cannot create those folders.
The easiest option would be to upgrade the VenusOS firmware. That will reset everything and force a reinstall from the driver.
If no new firmware version is available to upgrade too, switch to the backup firmware and then install a new version. This is all under the Settings->Firmware menu in your GX

**Reply by ghost** [reply]
Louis, it is working (see Device List.jpg and Device Battery.jpg)! In the /data/etc/dbus-serialbattery/utils.py the MAX_BATTERY_CURRENT = 50.0 but the Charge Current Limit (CCL) indicate 5.00A (see Battery 
![Battery Parameters](https://user-images.githubusercontent.com/24652007/174070385-f6864725-34a3-4158-b1de-b43ec2390922.jpg)
Paramers.jpg). By the way, I “had to upgrade” the Firmware. When last I looked, it was V2.66 and now it is V2.87. Thanks for your patience and assistance so far.
![Device List](https://user-images.githubusercontent.com/24652007/174070238-0b1702c9-ee2e-4cbb-8b91-ec6b11c8eb57.jpg)
![Device Battery](https://user-images.githubusercontent.com/24652007/174070265-957ecf4e-9633-4ace-a981-f7af0b96b8b0.jpg)

**Reply by Louisvdw** [reply]
Great!

You battery is full at 100% See the CCCM feature https://github.com/Louisvdw/dbus-serialbattery/wiki/Features#charge-current-control-management

**Reply by coca7** [reply]
> For some reason it cannot create those folders.
> The easiest option would be to upgrade the VenusOS firmware. That will reset everything and force a reinstall from the driver.
> If no new firmware version is available to upgrade too, switch to the backup firmware and then install a new version. This is all under the Settings->Firmware menu in your GX

I had this error on My Cerbo GX even on v2.7 and v2.73.  I just figured it was creating a folder that a non cerbo gx Venus OS install had, but a Cerbo GX did not.

**Reply by Louisvdw** [reply]
> I had this error on My Cerbo GX even on v2.7 and v2.73. I just figured it was creating a folder that a non cerbo gx Venus OS install had, but a Cerbo GX did not.

The builds before v0.12 used a symbolic link on those same paths that we want to create the folder on now. The install does remove those first, but it could be that those paths/links were created with a different user and that it does not have the rights to remove them now (or they exist already). A new firmware resets all that and get us on a known status to install on.

**Reply by ghost** [reply]
I now need to get that data to Telegraf and then Grafana so that I can see some neat looking graphs of that batteries, the mains power consumption, and voltages from the emonPi and then continue working on the python scripts to get the MicroCare MPPT to do the same.

**Reply by ghost** [reply]
Hi Louis, I am not sure whether this is right, but the SerialBattery system was showing 100% SOC which was incorrect since the SmartShunt was showing 78% SOC. The 78% was more reflective of the State of Charge. My question is: would this come right after a full-day’s use or is there something else that need to be adjusted.
![2022-06-17_10_23 Device List](https://user-images.githubusercontent.com/24652007/174280643-e6759128-d6b2-4098-a863-a31446bbcf2b.jpg)

**Reply by Louisvdw** [reply]
@bertiebez your BMS thinks your battery is at 100%, and so that is what the driver will display for you.
You will need to look at the settings in your BMS to callibrate it and make sure they match the battery. Most BMS will adjust their SOC in the first few (10 or so) cycles of your battery. They will use alarm inputs and cell voltage to adjust. This is normally when you have build a battery and the BMS would need to guess the starting point of the cells. You cells could be at 30% and the BMS would not know that, so it will use a few cycles to adjust. 

Each BMS is different. 
JBS will start ot 50% and then have a max adjust limit that it can adjust on each cycle.
JKBMS will start at 0%
I don't know what the Tian BMS will do.

From your battery voltage of 54.36V it does seem low to be at 100% (assuming that this is a 16 cell battery?) at 3.39V
Also remember that the 2 measurements will never be the same. They use different inputs, and so will have different results.
See https://github.com/Louisvdw/dbus-serialbattery/wiki/FAQ

**Reply by ghost** [reply]
In response to your last reply to my message I wish to add several points.
1.	Yesterday, I unplugged the SmartShunt hoping that the SerialBattery system will record the correct details. I am aware that SerialBattery system’s details will not be accurate, and it must go through several cycles, as was the case with the SmartShunt.
2.	After I unplugged the SmartShunt, the DC system power input (from the MPPT) on the Cerbo GX gave me a totally “inaccurate” reading (DC System). It is about half of what is in fact fed into the system (which is also the combined total of the battery plus AC loads). See the attached images
3.	On the device display, the battery voltage is correct with the SOC (100%) and Charging Current (5.00A) remaining unchanged.
Later today, I will use the software given to me before to download a fresh set of detail from the battery, but in the meantime, see the attached screen shot.
![Cerbo GX Main Display](https://user-images.githubusercontent.com/24652007/174425915-cf863904-77d1-4303-a49f-c24b91fb2daf.jpg)
![2022-06-18_08_10 Device List](https://user-images.githubusercontent.com/24652007/174425922-2c59658c-d6e6-4f7c-997b-e1863c1a96f8.jpg)
![2019-09-02 Bertie Battery #1](https://user-images.githubusercontent.com/24652007/174426009-01fa6b39-7888-4c31-934e-14d059958120.JPG)

**Reply by ghost** [reply]
Attached is a copy of the screenshot of the state of the battery at 09:01 this morning.
![Bertie Battery 2002-06-18_09-01](https://user-images.githubusercontent.com/24652007/174427141-8ca83d78-9f7a-4c23-848c-b31efe871cf2.jpg)

**Reply by Louisvdw** [reply]
Hi @bertiebez 
SOC is a difficult consept to understand sometimes. And as each device also define this as something different it makes it even more confusing for people. 
In this case the SOC values from your BMS is correct and your Smartshunt will be wrong. 

The Smartshunt only counts the power flowing in each direction. So if you have a 100Ah battery and you have used(discharge) 80Ah it will say your SOC is at 20%. If you add (charge) 70Ah it will say your battery SOC is 90%.

What happens with a BMS is that it does not only count the power flowing, but it also know about the status of each cell inside the battery. It can see that you have most cells at 3.3V, but then there are 2x 3.5V, 2x 3.6V and 1x 3.7V which will trigger the overcharge protections and it will stop the charging to protect your battery from exploding into a fireball. That means you cannot add any more power to the battery, so it is at 100% SOC (100% State Of Charge, cause you cannot charge it any more). So even if it has not reach full capacity for your battery, it is fully charged.

Some batteries will will also have SOH (State Of Health) which would be the difference between your capacity and the charge. (You will notice your show 43% SOH because of your issues).
What this show is that your cells are not ballanced. Either the cells themselve have been damanged, or the ballancer on your BMS is not working well. From what I heard this seem to be a common issue with the Tian BMS and Revov batteries.
The CCCM feature of the driver is to help give the BMS more time with reduced current when the battery is at this 100% SOC, but we reduce the current to 5A and most balancers are only 50mA-100mA or some more which means it can reduce the power in one problem cell by 100mA while the battery is charging at (5000 / 16 cells, or 312mA) and in your case the cells are too far for that little bit to help.
You will need to manually balance those cells first, and then CCCM might be able to help the Tian BMS to keep them in check.
The other option is to by an external balancer and disable the Tian BMS balancer, or replace the BMS complately.

**Reply by ghost** [reply]
Hi @bertiebez
SOC is a difficult consept to understand sometimes. And as each device also define this as something different it makes it even more confusing for people.
In this case the SOC values from your BMS is correct and your Smartshunt will be wrong.

The Smartshunt only counts the power flowing in each direction. So if you have a 100Ah battery and you have used(discharge) 80Ah it will say your SOC is at 20%. If you add (charge) 70Ah it will say your battery SOC is 90%.

What happens with a BMS is that it does not only count the power flowing, but it also know about the status of each cell inside the battery. It can see that you have most cells at 3.3V, but then there are 2x 3.5V, 2x 3.6V and 1x 3.7V which will trigger the overcharge protections and it will stop the charging to protect your battery from exploding into a fireball. That means you cannot add any more power to the battery, so it is at 100% SOC (100% State Of Charge, cause you cannot charge it any more). So even if it has not reach full capacity for your battery, it is fully charged.

Some batteries will will also have SOH (State Of Health) which would be the difference between your capacity and the charge. (You will notice your show 43% SOH because of your issues).
What this show is that your cells are not ballanced. Either the cells themselve have been damanged, or the ballancer on your BMS is not working well. From what I heard this seem to be a common issue with the Tian BMS and Revov batteries.
The CCCM feature of the driver is to help give the BMS more time with reduced current when the battery is at this 100% SOC, but we reduce the current to 5A and most balancers are only 50mA-100mA or some more which means it can reduce the power in one problem cell by 100mA while the battery is charging at (5000 / 16 cells, or 312mA) and in your case the cells are too far for that little bit to help.
You will need to manually balance those cells first, and then CCCM might be able to help the Tian BMS to keep them in check.
The other option is to by an external balancer and disable the Tian BMS balancer, or replace the BMS complately.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.

If you look at the last screenshot that I posted, you will notice that the State of Charge (SOC) at 09:01 on Saturday was in fact 70.48%. Perhaps one can say that it would be wrong, but I do however have a fair amount of trust in that figure since I have been watching the state of the batteries for three years. At some point I have turned off any means of charging the batteries so that I can get it to “trip” on low voltage. I then took the readings using the same software and it corresponds to the actuals.
I agree that the SmartShunt cannot give an accurate reading since it determines the State of Charge based on the current flowing into and out of the batteries. But the SOC on the SerialBattery is also not correct since it cannot stay put at 100%.
Referring once again to the screenshot taken on Saturday at 09:01, you will notice the current was 24.69 which was about the same as the battery charging current, which I read on the MPPT.
My assumption (and I may be wrong) is that there could be an issue with the protocols and thus the data displayed on the Cerbo GX.
I have extracted the data into a text file and perhaps you could have a look at what I have.
I don’t have another means of corresponding with you, so, the only way is through this platform.
![Bertie Battery 2002-06-18_09-01](https://user-images.githubusercontent.com/24652007/174489025-5ca5c64e-2ef8-4e0c-b353-3f61b09e2447.jpg)

**Comment by Louisvdw** [comment]
It is important to remember that the serialbattery only populate the battery details that it reads from your battery. If the values you see in your app is the same as what the driver display in your GX (or close as there might be a small time delay), then everything is working as it should. The driver does not change anything on the BMS/battery. It is a read only transaction.
The GX and the battery app show the same SOC from your screenshot. So that is working fine.

The issue where you are expecting your battery to slowly charge up to 100%, but it is jumping from around 70% to 100% is not a problem with the driver, but a problem with your battery. The driver is just showing you that this issue exist. The problem will stay even if you remove the driver.

It is jumping to 100% because the logic in the BMS is saying that we have an critical alarm and cannot charge the battery any more. 
Fix the imballance issue on your battery and the SOC will behave the way you expect it and you will be able to use the full capacity of your battery. If all your cells have a voltage of 3.45V +- 0.03 then you can say that the cells are balanced and at 100%SOC and 100% Capacity and your SOH will also be around 100% (it does not seem that the Tian BMS remember the SOH issues between cycles which is bizarre!)
![image](https://user-images.githubusercontent.com/26077665/174491386-d0a0975d-cbb5-47b8-885b-0ff2cfa47471.png)

If you want to read more about someone who had similar issues and how they solved it by top balancing the cells see the posts below (sorry it's about 250+ posts, but I don't want to repeat all that)
https://energytalk.co.za/t/lithium-cells-bulging-what-are-the-opinions/637

**Reply by ghost** [reply]
The Cerbo GX and the battery app show more-or-less the same SOC as for the SmartShunt. But as I said in my previous posting, I don’t believe that the SOC from the SmartShunt is accurate. However, the SOC from the SerialBattery (Revov) is even close to the same. In the images included with this posting you will notice that the “battery app” shows the SOC at 58.42% while the SmartShunt shows it as 63% and the SerialBattery (Revov) shows 100%. The current (13.28A) on the “battery app” also closely resemble the SmartShunt’s current (10.1A) which was taken at different times. Yes, it is about a minute apart, but I saw the variances as I was taking the snapshots.
By the way, these batteries are of a better quality and not the originals. I am not going to get into the details of what happened before, but the Revov replaced the first set after I kicked up a row showing them similar screenshots of exactly what you referred to in your last reply. They also replaced the BMS wih a newer version. The cell voltages were all over the show and could never balance properly. I downloaded the cell voltages and a lot more details on a regular basis until they realised that there was a problem. Yes, there could be problems with the batteries, but the Midpoint Deviation is 0.0% and the difference between the Lowest and Highest cell voltages are less that 0.26% which is 0.02 volts.
I looked through all the images of the info I took of the state of the batteries since it was replaced, and I could not find voltages as high as in the image attached to your reply. Yes, as I said in the previous paragraph, the voltages were all over the show.
I will analyse the file with the hex details I downloaded a little later to see if I can make sense of what was downloaded.
I wish to help you get this one right and would be of assistance as far as I can.
![Cergo GX Device List 2022-06-20_08-11](https://user-images.githubusercontent.com/24652007/174543499-27fa7933-bece-4ab7-ac2c-68a034b9e932.jpg)
![Cerbo GX Main Display 2022-06-20_08-10](https://user-images.githubusercontent.com/24652007/174543506-caef7b1e-4a4a-4024-b119-d4e201220b66.jpg)
![Baterry Info 2022-06-20_08-06](https://user-images.githubusercontent.com/24652007/174543507-a51f5cd0-dae1-4bc8-90ca-a3a3b60512e3.jpg)

**Reply by Louisvdw** [reply]
Hi

I have reread all your posts a few times and I see in the middle of one post you mention 
> it cannot stay put at 100%

Is that the problem you are trying to tell me? If that is the case then there might well be an issue. Can you show me the SOC% chart for a day from VRM please. Also please also check the log file if there are errors /data/log/dbus-serialbattery.ttyUSB0/current

Comments from some of the other things you mention:

The problem with real time data is that is changes in real time. So any screenshot that is a second apart can have many things changed during that time. That is where the graphs in Victrons VRM is such a great tool to find problems, as you can see the data over the timeline and also match and view data at a spesific date. For instance your battery could be in an overvoltage alarm state and then 2 seconds later some load will switch on and draws power immediately pulling down all your cells by 0.2V which removes that alarm. 

Also note that depending on the capabilities of your BMS it could only handle 1 serial connect at a time, or it could handle more. In your case I don't think it can as in your last screenshot it gets a Wrong Data  status which could mean if failed to get the data and you are looking at old values.
![image](https://user-images.githubusercontent.com/26077665/174657394-c18c4f8f-d44d-4b2e-9b2b-6f7c669ad2ee.png)


> and I could not find voltages as high as in the image attached to your reply

Just scroll up. That is a screenshot that I took of the images you posted earlier. It is the image just below the one where your SmartShunt is disconnected. That image show a battery with cells that is very unbalanced.

> Midpoint Deviation is 0.0% and the difference between the Lowest and Highest cell voltages are less that 0.26% which is 0.02 volts

When you measure this, was your battery at 100% SOC? Between 30% an 70% the voltages of lithium cells look the same. But if one cell is at 100% (3.45V) the other cells will then only be at 60% (still 3.3V). So you need to measure the cell deviation when the battery is full or depleted to see the real value. The Min/Max cell Voltage graph in VRM will show this over a day which is great to see this.

**Reply by ghost** [reply]
After plenty of searching I found the image you are referring to with the high voltage on cell 10. That image was of the “faulty” batteries which was since replaced. I included the image to show that SOC of the batteries as obtained through the application I often use to check on the state of the batteries, do change. I have used this application many times in the past following my experience with the original “poor quality batteries”. That is why I posted another image shortly after including that one (the one you are referring to) was posted.
As for the comment on the “Wrong Data status”, I may have unplugged the USB shortly before taking the screenshot or something like that. The Current and SOC are, as far as I am concerned, correct. It kept changing while I was busy taking preparing to take the screenshot.
As to the data in the log file, I can report that there are several warnings. From cell 1 to 16, it all show “low voltage”, but that is because of not being charged and a “high” power demand. I will include the text file with the data I extracted.
Aldo included are images of the SOC from the VRM plus one of an Excel graph (obtained from the Excel spreadsheet) with the SOC from the SerialBattery, SmartShunt and the “System overview” from midnight last night (00:00:00) until 07:51:39 this morning. In that you can see the 100% of serial 1 while series 2 and 3 goes down.
Perhaps you can, if you have it, send me a modpoll command line so that I can see if I can extract the data from the BMS.
[dbus-serialbattery_ttyUSB0.txt](https://github.com/Louisvdw/dbus-serialbattery/files/8946709/dbus-serialbattery_ttyUSB0.txt)
![SOC Excel Graph](https://user-images.githubusercontent.com/24652007/174741322-2e8e351f-ce89-4d15-ad0b-d2fabd01737a.jpg)
![SOC VRM Graph](https://user-images.githubusercontent.com/24652007/174741332-1ec454be-b8ee-4b19-9f4d-ce80630c0799.jpg)

**Reply by ghost** [reply]
In my search for possible answers to why the SOC on the Revov battery monitoring “stay put at 100%”, it appears as if that is still on the “to do list” or, as someone else put it, “I do need to get the SoC state calc done and added in”. so, would you kindly let me know when a new release of the particularly the revov.py becomes available.

**Reply by Louisvdw** [reply]
@bertiebez make sure you are subscribed to the Tian Power BMS issue to get all the notifications https://github.com/Louisvdw/dbus-serialbattery/issues/14
